File: solutions/fsharp/rational-numbers/1/RationalNumbers.fs
// For my Shree DR.MDD
module RationalNumbers

type Ratio = Ratio of int * int

let gcd a b =
    let rec loop x y =
        if x % y = 0 then y
        else loop y (x % y)
    if a > b then loop a b else loop b a

let reduce (Ratio (n, d)) =
    if n = 0 then Ratio (0, 1)
    else
        let nn, dd = if d < 0 then -n, -d else n, d
        let g = gcd nn dd
        Ratio (nn / g, dd / g)

let create a b = Ratio (a, b) |> reduce

let add (Ratio (n1, d1)) (Ratio (n2, d2)) =
    Ratio (n1 * d2 + n2 * d1, d1 * d2) |> reduce

let sub (Ratio (n1, d1)) (Ratio (n2, d2)) =
    Ratio (n1 * d2 - n2 * d1, d1 * d2) |> reduce

let mul (Ratio (n1, d1)) (Ratio (n2, d2)) =
    Ratio (n1 * n2, d1 * d2) |> reduce

let div (Ratio (n1, d1)) (Ratio (n2, d2)) =
    if n2 * d1 = 0 then failwith "Division by zero"
    else Ratio (n1 * d2, n2 * d1) |> reduce

let abs (Ratio (n, d)) =
    Ratio (abs n, abs d) |> reduce

let exprational p (Ratio (n, d)) =
    if p = 0 then Ratio (1, 1)
    elif n = 0 then Ratio (0, d)
    else
        let nn, dd, pp = if p > 0 then n, d, p else d, n, -p
        Ratio (pown nn pp, pown dd pp) |> reduce

let expreal (Ratio (n, d)) b =
    let root v deg = System.Math.Pow(float v, 1.0 / float deg)
    root (pown b n) d

File: solutions/lua/rational-numbers/1/rational-numbers.lua
-- Dedicated to Shree DR.MDD
local function gcd(x, y)
  x, y = math.abs(x), math.abs(y)
  while y ~= 0 do
    x, y = y, x % y
  end
  return x
end

local function normalize(r)
  local n, d = r[1], r[2]
  if d == 0 then return {n, d} end
  local g = gcd(n, d)
  n, d = n / g, d / g
  if d < 0 then n, d = -n, -d end
  return {n, d}
end

local function add(a, b)
  return normalize({a[1]*b[2] + b[1]*a[2], a[2]*b[2]})
end

local function sub(a, b)
  return normalize({a[1]*b[2] - b[1]*a[2], a[2]*b[2]})
end

local function mul(a, b)
  return normalize({a[1]*b[1], a[2]*b[2]})
end

local function div(a, b)
  return normalize({a[1]*b[2], a[2]*b[1]})
end

local function abs(r)
  return normalize({math.abs(r[1]), math.abs(r[2])})
end

local function pow_rat(r, p)
  if p >= 0 then
    return normalize({r[1]^p, r[2]^p})
  else
    local k = -p
    return normalize({r[2]^k, r[1]^k})
  end
end

local function pow_real(x, r)
  return (x ^ r[1]) ^ (1 / r[2])
end

return {
  add = add,
  subtract = sub,
  multiply = mul,
  divide = div,
  abs = abs,
  exp_rational = pow_rat,
  exp_real = pow_real,
  reduce = normalize
}
